xen/arch/x86/tpm.c

```diff
@@ -978,11 +971,150 @@ void tpm_hash_extend(unsigned loc, unsigned pcr, uint8_t *buf, unsigned size,
 #endif
 }
 
+static struct slr_entry_policy *tpm_get_policy(void)
+{
+#ifdef __EARLY_TPM__
+#define TPM_PANIC(...) return NULL;
+#else
+#define TPM_PANIC(args...) panic(args)
+#endif
+
+    struct txt_os_mle_data *os_mle;
+    struct slr_table *slrt;
+    struct slr_entry_policy *policy;
+    struct slr_policy_entry *policy_entry;
+
+    os_mle = txt_os_mle_data_start(__va(read_txt_reg(TXTCR_HEAP_BASE)));
+
+    slrt = __va(os_mle->slrt);
+    if ( slrt->magic != SLR_TABLE_MAGIC )
+        TPM_PANIC("SLRT has invalid magic value: %#08x!\n", slrt->magic);
+    /* XXX: are newer revisions allowed? */
+    if ( slrt->revision != SLR_TABLE_REVISION )
+        TPM_PANIC("SLRT is of unsupported revision: %#04x!\n", slrt->revision);
+    if ( slrt->architecture != SLR_INTEL_TXT )
+        TPM_PANIC("SLRT is for unexpected architecture: %#04x!\n",
+                  slrt->architecture);
+    if ( slrt->size > slrt->max_size )
+        TPM_PANIC("SLRT is larger than its max size: %#08x > %#08x!\n",
+                  slrt->size, slrt->max_size);
+    if ( slr_next_entry_by_tag(slrt, NULL, SLR_ENTRY_INTEL_INFO) == NULL )
+        TPM_PANIC("SLRT is missing Intel-specific information!\n");
+
+    policy = (struct slr_entry_policy *)
+        slr_next_entry_by_tag(slrt, NULL, SLR_ENTRY_DRTM_POLICY);
+    if (policy == NULL)
+        TPM_PANIC("SLRT is missing DRTM policy!\n");
+
+    /* SLRT policy entry must be the first one (so measuring order matches
+     * policy order) and point to SLRT. */
+    /* XXX: are newer revisions allowed? */
+    if ( policy->revision != SLR_POLICY_REVISION )
+        TPM_PANIC("DRTM policy in SLRT is of unsupported revision: %#04x!\n",
+                  slrt->revision);
+    if ( policy->nr_entries == 0 )
+        TPM_PANIC("DRTM policy in SLRT is empty!\n");
+
+    policy_entry = (struct slr_policy_entry *)
+        ((uint8_t *)policy + sizeof(*policy));
+    if ( policy_entry->entity_type != SLR_ET_SLRT )
+        TPM_PANIC("First entry of DRTM policy in SLRT is empty!\n");
```

Comment:
Message is inconsistent with check, maybe something like:

```suggestion
        TPM_PANIC("First entry of DRTM policy in SLRT has wrong type: %#04x!\n",
                  policy_entry->entity_type);
```

Reply:
Thanks: https://github.com/TrenchBoot/xen/compare/5ed262bc0df7696f824b5a547339a804fa9e5a52..39a1478db53029b133c75d4f15a0f5774ad92f8a#diff-5861d423276bd98c31fb2a501569110297cbe6488842e0359435a3af795c24aeR1037-R1038